rust/src/thread.rs

```diff
@@ -0,0 +1,126 @@
+use crate::syscall::*;
+use alloc::boxed::Box;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use core::cell::UnsafeCell;
+use spin::Mutex;
+pub struct Page(UnsafeCell<[u64; 512]>);
+use core::sync::atomic::AtomicBool;
+unsafe impl Sync for Page {}
+struct ThreadControlBlock<T> {
+    tid: Mutex<u32>,
+    ret: Mutex<Option<T>>, // not really lock
+    finished: AtomicBool,
+}
+
+impl<T> ThreadControlBlock<T> {
+    pub fn new_stub() -> Self {
+        ThreadControlBlock {
+            tid: Mutex::new(0),
+            ret: Mutex::new(None),
+            finished: AtomicBool::new(false),
+        }
+    }
+}
+
+struct ThreadRunner<T, F>
+where
+    F: FnOnce() -> T,
+    F: Send + 'static,
+    T: Send + 'static,
+{
+    tcb: Arc<ThreadControlBlock<T>>,
+    closure: Option<F>,
+    stack: *mut (),
+}
+pub trait ThreadRun {
+    fn run(&mut self) -> *mut ();
+}
+
+impl<T, F> ThreadRun for ThreadRunner<T, F>
+where
+    F: FnOnce() -> T,
+    F: Send + 'static,
+    T: Send + 'static,
+{
+    fn run(&mut self) -> *mut () {
+        let closure =
+            core::mem::replace(&mut self.closure, None).expect("you can't run a thread twice!");
+        let ret = (closure)();
+        let mut r = self.tcb.ret.lock();
+        *r = Some(ret);
+        drop(r);
+        self.tcb
+            .finished
+            .store(true, core::sync::atomic::Ordering::Relaxed);
```

Comment:
The memory order here should be `Release`, so that others can see all WRITEs before this point.